### PR TITLE
Fix concurrency issues for updating robot object

### LINKED ISSUE #1089: Resolve race condition in read and write to robot object
**Describe the bug**
When we read and write to the robot object we have a race condition in the sense that a write to the robot object in one thread may happen in-between a read and a write in a different thread. The example of this happening is whenever a mission finishes and the robot is set to `Status = Available`. If a telemetry event handler trigger simultaneously it may read the robot object and read the status as `Busy` and then write the status to `Busy` just after the `OnMissionUpdate` has just written the status to `Available`. As they are in different threads they do not share a context and will not be aware of the changes the other one has made.  

**To Reproduce**
Reproducing reliably can be difficult but by turning the publish intervals for telemetry in isar-robot to 1 second for all publishers you will see this behavior.

Once that has been done run a mission and observe what happens to the status whenever a mission finishes. It may also happen if you set the status directly through the API although it is not as often. 

**Expected behavior**
The robot object should be locked such that there is no possibility of reading and writing to it simultaneously across threads. We have to look at how we do read/writes throughout the system.

For example it may be smart to redesign a lot of our endpoints such that activities which require reading the robot don't read at the beginning and then update at the end but are smaller and closer to atomic actions. 

**Screenshots**

### PR DESCRIPTION
**Pull request by aestene**
Closes #1089

### CODE REVIEW

File: backend/api/EventHandlers/IsarConnectionEventHandler.cs [hunk 1]
```diff
@@ -94,7 +94,11 @@ private void AddTimerForRobot(IsarRobotHeartbeatMessage isarRobotHeartbeat, Robo
             timer.Start();
 
             if (_isarConnectionTimers.TryAdd(robot.IsarId, timer)) { _logger.LogInformation("Added new timer for ISAR '{IsarId}' ('{RobotName}')", robot.IsarId, robot.Name); }
-            else { _logger.LogWarning("Failed to add new timer for ISAR '{IsarId}' ('{RobotName})'", robot.IsarId, robot.Name); }
+            else
+            {
+                _logger.LogWarning("Failed to add new timer for ISAR '{IsarId}' ('{RobotName})'", robot.IsarId, robot.Name);
+                timer.Close();
```
Comment (oysand) [inline backend/api/EventHandlers/IsarConnectionEventHandler.cs:93]: How does the timer close for the if statement above

File: backend/api/Services/RobotService.cs [hunk 2]
```diff
@@ -88,20 +97,66 @@ public async Task<Robot> Update(Robot robot)
...
             var robot = await GetRobotsWithSubModels().FirstOrDefaultAsync(ev => ev.Id.Equals(id));
+            if (robot is null) { return null; }
+
+            _context.Robots.Remove(robot);
+            await _context.SaveChangesAsync();
+
+            return robot;
+        }
+
+        public async Task<Robot> Update(string id, string? name = null, string? isarId = null, RobotModel? robotModel = null, string? serialNumber = null,
```
Comment (oysand) [inline backend/api/Services/RobotService.cs:108]: How do we set a value back to null?

File: backend/api/appsettings.Development.json [hunk 3]
```diff
@@ -43,6 +43,6 @@
     "AccountName": "flotilladevsa"
   },
   "Database": {
-    "UseInMemoryDatabase": true
+    "UseInMemoryDatabase": false
```
Comment (oysand) [inline backend/api/appsettings.Development.json:46]: Set back to true

File: backend/api/Services/RobotService.cs [hunk 4]
```diff
@@ -88,20 +97,66 @@ public async Task<Robot> Update(Robot robot)
...
         }
 
-        private IQueryable<Robot> GetRobotsWithSubModels()
+        private IQueryable<Robot> GetRobotsWithSubModels(bool lockWithSemaphore = true)
         {
-            return _context.Robots.Include(r => r.VideoStreams).Include(r => r.Model).Include(r => r.CurrentArea);
+            IIncludableQueryable<Robot, Area?> robots;
+            if (lockWithSemaphore)
+            {
+                updateRobotSemaphore.WaitOne();
```
Comment (oysand) [inline backend/api/Services/RobotService.cs:152]: Consider removing this if it is not needed for the read step

File: backend/api/EventHandlers/IsarConnectionEventHandler.cs (hunk 1 above)
Comment (aestene) [reply backend/api/EventHandlers/IsarConnectionEventHandler.cs:100]: Closes on the timeout event.

File: backend/api/Services/RobotService.cs (hunk 2 above)
Comment (aestene) [reply backend/api/Services/RobotService.cs]: Removed that.

**Review by Christdej — APPROVED** [review]
LGTM

**Pull request merged**